IT BELONGS IN A MUSEUM: Adds Museum Trophy Cases, and Soapstone Messages

Iamgoofball opened the pull request:
# librarian before goofcode

![librarian2](https://cloud.githubusercontent.com/assets/4081722/21087561/d1272892-bfdb-11e6-94b3-3a8d3ec02947.jpg)

# librarian after goofcode

![rs_1024x759-140326100842-1024 2harrison-ford-indiana-jones](https://cloud.githubusercontent.com/assets/4081722/21087563/d60a26fc-bfdb-11e6-9e31-93cf9b01856e.jpg)

:cl: Iamgoofball
add: Added Chisels to the Librarian! Use it on floors or walls to leave messages for future crewmembers they can read with the examine verb.
add: Added Museum Trophy Cases to the library! Hit these with items, and type in a plaque, to turn these into trophies! Once it goes in, it stays forever in the pool. You can put an item back into the pool and off a display case by replacing it with a new item.
/:cl:

:cl: Kor
add: The Librarian now has access to the treasure hunter's hat/uniform/coat, as well as the curator's whip.
/:cl:

![dreamseeker_2016-12-11_19-41-30](https://cloud.githubusercontent.com/assets/4081722/21087580/ff80d936-bfdb-11e6-861e-a2655331296f.png)

![dreamseeker_2016-12-11_19-39-41](https://cloud.githubusercontent.com/assets/4081722/21087581/02bcf0d0-bfdc-11e6-8400-4fba92547cc9.png)

Commit:
IT BELONGS IN A MUSEUM

KorPhaeron commented:
I'm going to merge this when it's up to standard the idea is great I'm going to ignore all comments about "ruining the theme" of the least popular job

ChangelingRain commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,49 @@
 	user << "\n"
+
+/obj/item/weapon/chisel
+	name = "chisel"
```
a chisel, pah! true messages would use a soapstone!

Iamgoofball replied:
k

ChangelingRain commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,49 @@
+
+/obj/structure/chisel_message/Destroy()
+	SSpersistence.new_chisel_messages -= src
+	..()
```
return ..()

ChangelingRain commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,49 @@
+/obj/item/weapon/chisel/afterattack(atom/target, mob/user, proximity)
+	if(!proximity)
+		return
+	if(istype(target, /turf/open/floor) || istype(target, /turf/closed/wall))
```
also you'll need something to make sure chisel messages don't spawn on turfs that are.... not floors or walls

RandomMarine commented:
Should be able to unanchor and pull them around to try and desperately save the displays from a loose singulo or something.

francinum commented:
does this sanitize mind-containing items?

GunHog commented:
Someone has been playing Dark Souls...

Iamgoofball commented:
@francinum Variables aren't kept
@RandomMarine They're indestructible just like your virginity

ChangelingRain commented:
Chisel messages also need to record what station they're on, so if at all possible they're at least sort of relevant instead of "these hallways are long as all fuck" on ministation, from dream

RandomMarine commented:
If they're somehow singulo-proof then 👎

KorPhaeron commented:
Reminder to 

-Limit placing items to the librarian somehow (ID lock would be okay, but still easily bypassed)

-A limit to the number of items you can add each round to the pool to prevent it being flooded

-Rename the Librarian something like Curator or Archivist

optimumtact commented on code/game/objects/structures/displaycase.dm:
```diff
@@ -114,7 +126,30 @@
+		if(is_type_in_typecache(W, blacklisted_cargo_types))
+			user << "You think putting [W] in would be a bad idea."
+			return
+		var/chosen_plaque = stripped_input(user, "What would you like the plaque to say?", "Trophy Plaque")
```
this is a sleep, check the item is not qdeleted and or already has an item inside it after this.

Iamgoofball replied:
aight

optimumtact commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,49 @@
+	if(istype(target, /turf/open/floor) || istype(target, /turf/closed/wall))
+		var/already_message = locate(/obj/structure/chisel_message) in target
+		if(!already_message)
+			var/message = stripped_input(user, "What would you like to engrave?", "Chisel Message")
```
IT'S A SLEEP BOB

optimumtact commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,49 @@
+			if(!message)
+				user << "You decide not to chisel anything."
+				return
+			if(do_mob(user, user, 10))
```
it's a SLEEP BOB

ExcessiveUseOfCobblestone commented:
@optimumtact can you explain what a sleep is and how to "fix" the sleep?

GunHog commented:
They should NOT be indestructible! When the case is destroyed, simply say that the item is old and crumbled to dust or something; Delete the item when the case is broken.

This is to create the situation where a greyshirt can break in and smash all the cases to start a fight with the librarian!

Shadowlight213 commented:
Is there a limit or something to this? Also how often can something be added per round?
I don't want some memer to add everything on the station to it to the point where loading the list lags the server.

ExcessiveUseOfCobblestone commented:
are the messages suppose to be a DS ref

PKPenguin321 commented:
librarian should actually get a whip

KorPhaeron commented:
@PKPenguin321 

Patience

Iamgoofball commented:
@ExcessiveUseOfCobblestone Great Chest Ahead!

ExcessiveUseOfCobblestone commented:
![meme](https://i.imgur.com/7NrELYn.jpg)

optimumtact commented:
@ExcessiveUseOfCobblestone when I say sleep I actually mean that the method your calling can cause the current proc to block and stop executing, allowing other game logic to run in the meantime.

What it means practically is that when your code in that method starts to execute again the state of the underlying object may have changed (from some other event, i.e the object got blown up)

So you need to redo any sanity checks for state you have, as well as check the object is not qdeleted.

dannno commented:
cool idea

ExcessiveUseOfCobblestone commented:
![chisel.png](http://i.imgur.com/BDLUpXQ.png)

Commit:
soapstone + fixes + map changes

AnturK commented:
I'm ambivalent about this, librarian needs some love but persistence features are awful.

Iamgoofball commented:
@anturk It's all 100% cosmetic and flavor stuff.

AnturK commented:
Oh you can't take them out ? Sounds good then.

Iamgoofball commented:
Yup, once it goes in it never comes out.

dannno commented:
I assume there's a blacklist for things like the disk?

Iamgoofball commented:
Yup

PKPenguin321 commented:
maybe hold off on the disarm whip
or make it just not disarm
or make it only have a chance to disarm and slightly less damage
or make it do very little damage and disarm

Iamgoofball commented:
I literally give zero fucks about balance meme and will be ignoring it completely.

jughu commented:
I cant wait to see shitcueity chiseled in the brigs everywhere

HankMoodyExperience commented:
>not being able to construct display cases;
>cant move them around
>indestructible

m u h i n d e s t r u c t i b l e e x p l o i t s
make the cases appear roundstart for the librarian, as constructible, destroyable things, of like wood and glass or something.

and limit chisel use so some memer doesnt literally plaster an entire station with "nigger" carvings

KorPhaeron commented:
Display case blocks a vent on meta

There needs to be some way to remove items from the pool both in case someone fills the entire thing with 1k variants of wgw and to stop the pool growing forever. Maybe have the case destructible and lose the item inside if its blown up

Still need a limit on items added per round

Iamgoofball commented:
You can't add the same item over and over again in one round
I'll make cases explode
You can smash messages to erase them from the pool, I'll lower their health
Limiting items added per round kind of sucks but I guess I can look into it

KorPhaeron commented:
Merge my PR

KorPhaeron committed:
Dicks out for harambe

* Chaplain costumes and theater masks (#22109)

Ah hah hah ha! Ooh! Majestic! A hunter is a hunter, even in a dream. But, alas, not too fast! The nightmare swirls and churns unending!

* Automatic changelog generation for PR #22109 [ci skip]

* Curator Items

* Spawns in his bookbag because where else would this stuff go

* Revert "Spawns in his bookbag because where else would this stuff go"

This reverts commit c3a8613b01a8ac99b0d44e7eeb2e69fa8ca5801c.

YakumoChen reviewed (commented):
>Librarian gets free security armor and a guaranteed disarm now
>Implying that powergaming tiders won't use this every shift for the free armor

KorPhaeron commented:
Just like people who play CMO every round for the free space suit and infinite stun right?

RandomMarine commented:
Suggesting a rework.

Make each display start out empty to begin with. When an item is first added, the case remains unlocked and the item easily retrievable. No blacklist necessary.

If the item remains inside the display case when the round ends, the item's preserved inside and the case is locked the following rounds, impossible to retrieve what's inside.

Each display has a keypad with a code shared between them that's randomized each round. Librarian spawns with a paper with that code. When the code's punched in, the item in the display is destroyed and it's unlocked, allowing something new to be placed inside.

Finally, no big pool of items that will get bloated with WGW papers or junk, what's put in each individual display stays until it's destroyed.

YakumoChen commented:
CMO is a head of staff, though. Librarian is just another assistant with a fancier name.

Iamgoofball commented:
@YakumoChen I don't care, find somewhere else to whine.

Iamgoofball commented:
@RandomMarine I'll think about it.

optimumtact commented:
Great job kor, you ruined the PR

coiax commented on code/game/machinery/vending.dm:
```diff
@@ -911,7 +911,7 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 					/obj/item/clothing/suit/wizrobe/marisa/fake = 1,/obj/item/clothing/under/sundress = 1,/obj/item/clothing/head/witchwig = 1,/obj/item/weapon/staff/broom = 1,
 					/obj/item/clothing/suit/wizrobe/fake = 1,/obj/item/clothing/head/wizard/fake = 1,/obj/item/weapon/staff = 3,/obj/item/clothing/mask/gas/sexyclown = 1,
-					/obj/item/clothing/under/rank/clown/sexy = 1,/obj/item/clothing/mask/gas/sexymime = 1,/obj/item/clothing/under/sexymime = 1,/obj/item/clothing/suit/apron/overalls = 1,
+					/obj/item/clothing/under/rank/clown/sexy = 1,/obj/item/clothing/mask/gas/sexymime = 1,/obj/item/clothing/under/sexymime = 1,/obj/item/clothing/mask/rat/bat = 1,/obj/item/clothing/mask/rat/bee = 1,/obj/item/clothing/mask/rat/bear = 1,/obj/item/clothing/mask/rat/raven = 1,/obj/item/clothing/mask/rat/jackal = 1,/obj/item/clothing/mask/rat/fox = 1,/obj/item/clothing/mask/rat/tribal = 1,/obj/item ...
```
STEALTH CHANGES!?!?!?

coiax commented:
- I think you should make chisel messages in Box Station only appear on Box Station and so on, otherwise it loses all meaning.
- You should add (real world) dates and times of when the trophies were added and the character name, along with storing (internally) the ckey of who added it, for subsequent justice if they added the Bee Movie script and/or Nazis.
- The trophy pool will get very large, I think you should maybe limit to only one new trophy a round for sanity's sake. (So, assuming a librarian always trophies and 1 hour rounds, that's 168+ trophies a week). Bonus points if librarians can vote up trophies that they approve of, (COMMEND ME).
- Admins will need some method of viewing the trophy pool and PURGING AND PUNISHING.

Otherwise, :+1:x:100:, this is exactly the sort of fun shit that is cool.

FantasticFwoosh commented:
**_TAKE ALL MY MONEY._**

Though i sort of wish the curator role wasn't so heavily attributed to the librarian/archeologist and more a thing anybody could take up. **Using the the chisel for Goof-Archeology when?**

> this pr comments are starting to get full so put additional non-coding feedback here please on this forum thread https://tgstation13.org/phpBB/viewtopic.php?f=10&t=8849

AnturK commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,53 @@
 	user << "\n"
+
+/obj/item/weapon/chisel
+	name = "soapstone"
```
Give this a proper name instead of forced reference that makes no sense in context.

Iamgoofball replied:
I was told by a maintainer to make it a soapstone

Iamgoofball replied:
I will

optimumtact replied:
please fix the path name as well

Iamgoofball commented:
>I think you should make chisel messages in Box Station only appear on Box Station and so on, otherwise it loses all meaning.

Already done

AnturK commented on code/modules/library/lib_items.dm:
```diff
@@ -316,3 +316,53 @@
+				user << "You decide not to chisel anything."
+				return
+			if(do_mob(user, user, 10))
+				if(!target.Adjacent(user) && !locate(/obj/structure/chisel_message) in target)
```
target instead of !target

Iamgoofball replied:
Fuck

AnturK commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -65,6 +91,25 @@ var/datum/subsystem/persistence/SSpersistence
 	return 1
 
+/datum/subsystem/persistence/proc/PlaceChiselMessage(list/expanded_old_chisels)
+	for(var/chisel_string in expanded_old_chisels)
```
No cap on this ? Station will be one giant scribble.

AnturK commented on code/game/objects/structures/displaycase.dm:
```diff
@@ -114,7 +127,33 @@
+			added_roundstart = 0
+			user << "You insert [W] into the case."
+			types_added_this_shift += W.type
+			W.loc = src
```
forceMove

AnturK commented on code/game/objects/structures/displaycase.dm:
```diff
@@ -251,3 +290,12 @@
+	obj_integrity = INFINITY // no
+	max_integrity = INFINITY
+	start_showpiece_type = null
+	req_access = list(access_cent_bar)
```
cent_bar ?

Iamgoofball replied:
Wanted to make sure that you couldn't open it

Shadowlight213 replied:
Should double check that the centcom official/ert doesn't have this.
Actually, what would even happen if you removed the item? Would it not go back in the pool?

Iamgoofball replied:
It would let you get items from other rounds

Iamgoofball replied:
ale mao

AnturK commented on code/modules/clothing/glasses/glasses.dm:
```diff
@@ -286,7 +286,7 @@
 /obj/item/clothing/glasses/red
 	name = "red glasses"
-	desc = "A sweet pair of red shades."
+	desc = "Hey, you're looking good, senpai!"
```
???

Iamgoofball replied:
KORRRRRRRRR

KorPhaeron replied:
Wasnt me

KorPhaeron replied:
https://github.com/tgstation/tgstation/pull/22109/files

It IS in there, not sure why it shows as a difference

optimumtact replied:
did you squash merge kor's pr?

Iamgoofball replied:
yes

optimumtact replied:
Probably why, especially if it had a master merge

Iamgoofball replied:
Ale Mao

Haevacht commented:
As long as it's still a chisel and not the soapstone.

Iamgoofball commented:
It's staying a soapstone it's literally a dark souls joke

Haevacht commented:
why

karlnp commented:
Are they singulo-proof?

Jazaen commented:
You should add some way of removing items from the pool, in case some asshole just labels a bunch of floor tiles idiotic stuff and throws them into the case to make sure nothing interesting ever appears.

ExcessiveUseOfCobblestone commented:
I want people to go explore space now so they can put erotic messages below the Venus statues.

Dorsisdwarf commented:
I love this except for the librarian getting armor for no reason.

Its a crumpled archaeology hat not combat gear you dork he doesnt need armor if he whips the gun out of the syndicate thief's hand

dannno commented:
it does seem pretty stupid for it to have armor for no actual reason
it's just a hat and some clothes

Shadowlight213 commented:
Don't forget the force 15 weapon that makes the target drop all held items on hit!
The officer cannot arrest if you disable his hand!

KorPhaeron commented:
My thought process behind giving him basic armour was that he'd have a little protection if he ran around in lavaland or something looking for artifacts

KorPhaeron commented:
>complaining about force 15 when there is a public lathe on meta that produces infinite force 15 weapons

The disarm part is the good part, and it's still worse than a baton.

dannno commented:
the whip doesn't seem like a problem at all to me
it would be better if it was a 1 tile reach weapon with disarm that did less damage if adjacent to the target, but that's not a thing (yet)

Shadowlight213 commented:
The weapon is already really good. Useful on crew more than lavaland though.
The librarian actually going to lavaland is unlikely unless you add mining access to him.
The armor isn't needed.

Iamgoofball commented:
~~literally don't care~~

Dorsisdwarf commented:
An archaeologist who needs armor to rescue totems belongs in a museum

jughu commented:
@Iamgoofball whip better be a ranged weapon for 2 tiles

TehZombehz commented:
What, nobody made the obvious 'try tongue but hole' dark souls soapstone reference? Can't wait till this is testmerged and ~~get to play archeologist as librarian~~  ~~roll librarian to find all the tools were simply stolen by quite literally anybody and find the actual librarian job isn't needed for any of this~~ have fun.

kevinz000 commented on code/game/objects/items/weapons/melee/misc.dm:
```diff
@@ -261,3 +261,21 @@
 /obj/item/weapon/melee/supermatter_sword/add_blood(list/blood_dna)
 	return 0
+
+/obj/item/weapon/melee/curator_whip
```
Seems VERY powerful.

Armhulen commented:
goofball PLEASE finish this, just one last push until it's done

optimumtact commented:
Extract soapstoan/chisel into it's own pr

extract display cases into it's own pr.

use json for the persistence.

Thanks

Iamgoofball commented:
@optimumtact 
no
no
if i can get it working I will

optimumtact commented:
If you don't seperate them you don't get a merged pr.

This isn't a request, it's an order

Iamgoofball commented:
They're both for the librarian only though
